src/Compilers/Core/Portable/PEWriter/Types.cs

```diff
@@ -253,6 +253,9 @@ internal interface INamedTypeReference : ITypeReference, INamedEntity
         /// If true, the persisted type name is mangled by appending "`n" where n is the number of type parameters, if the number of type parameters is greater than 0.
         /// </summary>
         bool MangleName { get; }
+
+        /// <summary>Indicates that the type is scoped to the file it is declared in. Used as a prefix for the metadata name.</summary>
+        string? AssociatedFileIdentifier { get; }
```

Comment:
>AssociatedFileIdentifier [](http://example.com/codeflow?start=16&length=24)

It feels like too many language concepts is getting leaked to the common layer. We probably should investigate if we can get rid of the ```MangleName``` property instead, and have implementation to provide "true" metadata name through the ```Name``` property.

Reply:
I think we should try and at least push the C#-specific bits into the C# layer. I attempted this today but it got out of hand a little bit. Will take another look tomorrow.

I think there's a wider ranging impact to actually changing `INamedEntity.Name` to return the metadata name which might be more appropriate as a design debt item.